File: stvs/src/test/java/edu/kit/iti/formal/stvs/model/table/problems/CellTypeProblemTest.java
package edu.kit.iti.formal.stvs.model.table.problems;

import edu.kit.iti.formal.stvs.model.expressions.*;
import edu.kit.iti.formal.stvs.model.table.ConstraintCell;
import org.junit.Before;
import org.junit.Test;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static edu.kit.iti.formal.stvs.model.expressions.SimpleExpressions.and;
import static edu.kit.iti.formal.stvs.model.expressions.SimpleExpressions.literal;
import static org.hamcrest.CoreMatchers.instanceOf;
import static org.junit.Assert.*;

/**
 * Created by bal on 26.02.17.
 */
public class CellTypeProblemTest {

  private TypeCheckException typeCheckEx;
  private CellTypeProblem problem;

  @Before
  public void setUp() {
    typeCheckEx = new TypeCheckException(new BinaryFunctionExpr(BinaryFunctionExpr.Op.AND, new
        LiteralExpr(new ValueInt(2)), new LiteralExpr(ValueBool.TRUE)), "Expected type \"BOOL\","
        + "but got \"INT\"");
    problem = new CellTypeProblem(typeCheckEx, "A", 4);
  }

  @Test
  public void tryTypeCheckCellExpression() throws Exception {
    Map<String, Type> typeMap = new HashMap<>();
    typeMap.put("A", TypeInt.INT);
    typeMap.put("B", TypeBool.BOOL);
    TypeChecker typeChecker = new TypeChecker(typeMap);
    Expression problematicCell = and(literal(2), literal(true));
    try {
      CellTypeProblem.tryTypeCheckCellExpression(
          typeChecker, "A", 4, problematicCell);
    } catch (Exception exc) {
      assertThat(exc, instanceOf(CellTypeProblem.class));
      assertEquals(problem, exc);
    }
  }

  @Test
  public void getTypeCheckException() throws Exception {
    assertEquals(typeCheckEx, problem.getTypeCheckException());
  }

  @Test
  public void testEquals() throws Exception {
    CellTypeProblem identical = new CellTypeProblem(typeCheckEx, "A", 4);
    assertEquals(problem, identical);
    assertNotEquals(problem, null);
    assertEquals(problem, problem);
    CellTypeProblem notIdentical = new CellTypeProblem(typeCheckEx, "A", 5);
    assertNotEquals(notIdentical, problem);
  }

  @Test
  public void testHashCode() throws Exception {
    CellTypeProblem identical = new CellTypeProblem(typeCheckEx, "A", 4);
    assertEquals(problem.hashCode(), identical.hashCode());
    assertEquals(problem.hashCode(), problem.hashCode());
    CellTypeProblem notIdentical = new CellTypeProblem(typeCheckEx, "A", 5);
    assertNotEquals(notIdentical.hashCode(), problem.hashCode());
  }

}

File: stvs/src/test/kotlin/edu/kit/iti/formal/stvs/model/table/problems/CellTypeProblemTest.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.model.table.problems

import edu.kit.iti.formal.stvs.model.expressions.*
import edu.kit.iti.formal.stvs.model.table.problems.CellTypeProblem.Companion.tryTypeCheckCellExpression
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

/**
 * Created by bal on 26.02.17.
 */
class CellTypeProblemTest {
    private var typeCheckEx: TypeCheckException? = null
    private var problem: CellTypeProblem? = null

    @BeforeEach
    fun setUp() {
        typeCheckEx = TypeCheckException(
            BinaryFunctionExpr(BinaryFunctionExpr.Op.AND, LiteralExpr(ValueInt(2)), LiteralExpr(ValueBool.TRUE)),
            "Expected type \"BOOL\"," +
                "but got \"INT\"",
        )
        problem = CellTypeProblem(typeCheckEx!!, "A", 4)
    }

    @Test
    @Throws(Exception::class)
    fun tryTypeCheckCellExpression() {
        val typeMap = mapOf("A" to TypeInt, "B" to TypeBool)
        val typeChecker = TypeChecker(typeMap)
        val problematicCell = SimpleExpressions.and(SimpleExpressions.literal(2), SimpleExpressions.literal(true))
        try {
            tryTypeCheckCellExpression(
                typeChecker,
                "A",
                4,
                problematicCell,
            )
        } catch (exc: SpecProblemException) {
            Assertions.assertEquals(problem, exc.problem)
        }
    }

    @Test
    fun typeCheckException() {
        Assertions.assertEquals(typeCheckEx, problem!!.exception)
    }

    @Test
    @Throws(Exception::class)
    fun testEquals() {
        val identical = CellTypeProblem(typeCheckEx!!, "A", 4)
        Assertions.assertEquals(problem, identical)
        Assertions.assertNotEquals(problem, null)
        Assertions.assertEquals(problem, problem)
        val notIdentical = CellTypeProblem(typeCheckEx!!, "A", 5)
        Assertions.assertNotEquals(notIdentical, problem)
    }

    @Test
    @Throws(Exception::class)
    fun testHashCode() {
        val identical = CellTypeProblem(typeCheckEx!!, "A", 4)
        Assertions.assertEquals(problem.hashCode().toLong(), identical.hashCode().toLong())
        Assertions.assertEquals(problem.hashCode().toLong(), problem.hashCode().toLong())
        val notIdentical = CellTypeProblem(typeCheckEx!!, "A", 5)
        Assertions.assertNotEquals(notIdentical.hashCode().toLong(), problem.hashCode().toLong())
    }
}